### PR TITLE
[lldb-dap] Support finding the lldb-dap binary

### PR DESCRIPTION
**Pull request by JDevlieghere**
Support finding the lldb-dap binary with `xcrun` on Darwin or in PATH on all other platforms.

Unfortunately, this PR is larger than I would like because it removes the `lldbDapOptions`.  I believe these options are not necessary, and as previously implemented, they caused a spurious warning with this change. The problem was that the options were created before the custom factory. By moving the creation logic into the factory, we make sure it's only called after the factory has been registered. The upside is that this simplifies the code and removes a level of indirection.

### CODE REVIEW

File: lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts [hunk 1]
```diff
@@ -28,6 +23,70 @@ export class LLDBDapDescriptorFactory
     return true;
   }
 
+  static async findDAPExecutable(): Promise<string | undefined> {
+    let executable = "lldb-dap";
+    if (process.platform === "win32") {
+      executable = "lldb-dap.exe";
+    }
```
Comment (walter-erquinigo) [inline lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:30]: just use the ternary comparator

File: lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts [hunk 2]
```diff
@@ -28,6 +23,70 @@ export class LLDBDapDescriptorFactory
...
+
+    // Prefer lldb-dap from Xcode on Darwin.
+    if (process.platform === "darwin") {
+      try {
+        const exec = util.promisify(require("child_process").execFile);
```
Comment (walter-erquinigo) [inline lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:35]: don't use require. Just import child process in the regular way, and then pass it to promisify. That way the typechecker will work correctly.

File: lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts [hunk 3]
```diff
@@ -28,6 +23,70 @@ export class LLDBDapDescriptorFactory
...
+        if (stdout) {
+          return stdout.toString().trimEnd();
+        }
+      } catch (error) {}
+    }
+
+    // Find lldb-dap in the user's path.
+    let env_path =
+      process.env["PATH"] ||
+      (process.platform === "win32" ? process.env["Path"] : null);
```
Comment (walter-erquinigo) [inline lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:49]:
```suggestion
      process.platform === "win32" ? process.env["Path"] : process.env["PATH"]
```

File: lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts [hunk 4]
```diff
@@ -28,6 +23,70 @@ export class LLDBDapDescriptorFactory
     return true;
   }
 
+  static async findDAPExecutable(): Promise<string | undefined> {
```
Comment (walter-erquinigo) [inline lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:26]: This function is really not complex but a bit long. Can you split it into smaller chunks?

**Review by walter-erquinigo — APPROVED** [review]
beautiful

File: lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts [hunk 5]
```diff
@@ -8,15 +11,7 @@ import { LLDBDapOptions } from "./types";
...
-  private lldbDapOptions: LLDBDapOptions;
-
-  constructor(lldbDapOptions: LLDBDapOptions) {
-    this.lldbDapOptions = lldbDapOptions;
-  }
-
-  static async isValidDebugAdapterPath(
-    pathUri: vscode.Uri,
-  ): Promise<Boolean> {
+  static async isValidFile(pathUri: vscode.Uri): Promise<Boolean> {
```
Comment (ashgti) [inline lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:14]: Should this just be a function outside of the class?

File: lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts [hunk 6]
```diff
@@ -8,15 +11,7 @@ import { LLDBDapOptions } from "./types";
...
     try {
```
Comment (ashgti) [inline lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:15]: We may want to validate that the file is executable. It looks like the VS Code FS API doesn't have that info but we could use the node FS instead, something like:
```
import { access, constants } from 'node:fs/promises';

try {
  await access('<path>', constants.X_OK);
  console.log('can execute');
} catch {
  console.error('cannot execute');
} 
```

File: lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts (hunk 2 above)
Comment (ashgti) [inline lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:35]: Should this be moved to the top of the file? Its never modified after creation.

File: lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts (hunk 5 above)
Comment (JDevlieghere) [reply lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:14]: We're also calling this from `extension.ts`. I guess we could export the function but that doesn't seem much better than leaving it a static method?

Comment (ashgti) [reply lldb/tools/lldb-dap/src-ts/debug-adapter-factory.ts:14]: I don't think this is doing anything specific for `LLDBDapDescriptorFactory` and a top level `isValidExe` or `isExecutable` is straight forward to reason vs `LLDBDapDescriptorFactory.isValidFile`.

**Review by ashgti — APPROVED** [review]
Super useful, especially on macOS with Xcode installed

**Pull request merged**